File: BST.java
import java.util.Scanner;

class Node {
    int data;
    Node left, right;

    public Node(int item) {
        data = item;
        left = right = null;
    }
}

class BSTt {
    Node root;

    BSTt() {
        root = null;
    }

    void insert(int key) {
        root = insertRec(root, key);
    }

    Node insertRec(Node root, int key) {
        if (root == null) {
            root = new Node(key);
            return root;
        }

        if (key < root.data)
            root.left = insertRec(root.left, key);
        else if (key > root.data)
            root.right = insertRec(root.right, key);

        return root;
    }

    void preorder() {
        preorderRec(root);
        System.out.println();
    }

    void preorderRec(Node root) {
        if (root != null) {
            System.out.print(root.data + " ");
            preorderRec(root.left);
            preorderRec(root.right);
        }
    }

    void inorder() {
        inorderRec(root);
        System.out.println();
    }

    void inorderRec(Node root) {
        if (root != null) {
            inorderRec(root.left);
            System.out.print(root.data + " ");
            inorderRec(root.right);
        }
    }

    void postorder() {
        postorderRec(root);
        System.out.println();
    }

    void postorderRec(Node root) {
        if (root != null) {
            postorderRec(root.left);
            postorderRec(root.right);
            System.out.print(root.data + " ");
        }
    }
}

public class BST {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        BSTt tree = new BSTt();

        System.out.println("Enter number of nodes to insert:");
        int n = sc.nextInt();

        System.out.println("Enter " + n + " values to insert into the BST:");
        for (int i = 0; i < n; i++) {
            int val = sc.nextInt();
            tree.insert(val);
        }

        System.out.println("Preorder traversal:");
        tree.preorder();

        System.out.println("Inorder traversal:");
        tree.inorder();

        System.out.println("Postorder traversal:");
        tree.postorder();

        sc.close();
    }
}

File: BST.py
# Node class for BST
class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

# Insert function
def insert(root, key):
    if root is None:
        return Node(key)
    else:
        if key < root.key:
            root.left = insert(root.left, key)
        else:
            root.right = insert(root.right, key)
    return root

# Inorder Traversal (Left → Root → Right)
def inorder(root):
    if root:
        inorder(root.left)
        print(root.key, end=" ")
        inorder(root.right)

# Preorder Traversal (Root → Left → Right)
def preorder(root):
    if root:
        print(root.key, end=" ")
        preorder(root.left)
        preorder(root.right)

# Postorder Traversal (Left → Right → Root)
def postorder(root):
    if root:
        postorder(root.left)
        postorder(root.right)
        print(root.key, end=" ")

# --- Main Program ---
if __name__ == "__main__":
    n = int(input("Enter number of nodes: "))
    root = None

    print("Enter the node values:")
    for _ in range(n):
        value = int(input())
        root = insert(root, value)

    print("\nInorder Traversal: ")
    inorder(root)

    print("\nPreorder Traversal: ")
    preorder(root)

    print("\nPostorder Traversal: ")
    postorder(root)


# Enter number of nodes: 7
# Enter the node values:
# 50
# 30
# 20
# 40
# 70
# 60
# 80

# Inorder Traversal: 
# 20 30 40 50 60 70 80
# Preorder Traversal: 
# 50 30 20 40 70 60 80
# Postorder Traversal: 
# 20 40 30 60 80 70 50

